### PR TITLE
[30577] Make work packages module to BCF module when activated

### PR DESCRIPTION
**Pull request by HDinger**
When activated the "Work Package" module shall be renamed to "BCF". In addition to the existing WP functionality, the `import`/ `export` button of the BCF module shall be added to the toolbar.

### ToDo
- [x] Rename WP Module when BCF is activated
- [x] Show import button in the toolbar, next to the create button (and make it work).
- [x] Show export button in the toolbar (and make it work)
- [x] Add a default BCF query, when the module is activated, which has the card view as default
- [x] Take care of correct redirections
  * Cancel of the import leads back to the list
  * A successful import leads to the "summary" page
  * The button on the summary page leads to the list 
- [x] Change the icon

https://community.openproject.com/projects/openproject/work_packages/30577/activity
https://community.openproject.com/projects/openproject/work_packages/30618/activity

### CODE REVIEW

**Comment by wielinde** [comment]
Hey @HDinger, thank you very much!! That PR will improve the UX of the BIM edition a lot.

Before starting the code review I opted for a quick "click through". Let me share some findings with you:

- During an import, when asked how to proceed with unknown entities such as Types, Statuses, Users and Priorities the Cancel-Buttons redirect to old controller index. It should redirect to the work packages module. Best would be to redirect back to the query where the user came from. I am not sure if that is easy to achieve. If not then simply go back to the work packages module's default view.
- After importing your PR directly redirects to the work package module. That happens one step too early. Before there was a view that showed the results of that import. And in case there were error messages associated with WP/BCF-issue, would get shown there. That view has a pretty hard to reach "Back to list" button which probably should be at the top of the page too and link back to work package module.
- Now seeing the buttons in the WP module's toolbar I think it is save to simply drop the "BCF-XML" from the export button label. I know I said something else before.

**Comment by HDinger** [comment]
@wielinde I included your feedback. For pragmatic reasons I have not implemented the redirect to the original query. Instead, it always returns to the default BCF query. I think the perfect redirect can still be done in a later step.

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -308,6 +308,8 @@ def body_css_classes
 
     css << "ee-banners-#{EnterpriseToken.show_banners? ? 'visible' : 'hidden'}"
 
+    css << "bcf-#{@project&.module_enabled?(:bcf) ? 'activated' : 'deactivated'}"
+
```
Comment (wielinde) [inline app/helpers/application_helper.rb]: Any chance to move that logic to the BCF module? For example by patching the method?

File: app/helpers/work_packages_filter_helper.rb [hunk 2]
```diff
@@ -214,6 +214,20 @@ def project_version_property_path(version, property_name, property_id, options =
     project_work_packages_with_query_path(version.project, query, options)
   end
 
+  def project_work_packages_bcf_issues_path(project)
```
Comment (wielinde) [inline app/helpers/work_packages_filter_helper.rb]: Same as above. Any chance we can move that out of the core to maintain the separation of concerns?

File: lib/redmine/plugin.rb [hunk 3]
```diff
@@ -273,6 +273,16 @@ def hide_menu_item(menu_name, item, hide_if: -> (*) { true })
       end
     end
 
+    def rename_menu_item(menu_name, item, options)
+      Redmine::MenuManager.map(menu_name) do |menu|
+        menu_item = menu.find(item)
+        menu_item.caption = options[:caption]
+        menu_item.icon = options[:icon]
+        menu_item.badge = options[:badge]
+
```
Comment (wielinde) [inline lib/redmine/plugin.rb]: obsolete empty line

File: modules/bcf/lib/open_project/bcf/engine.rb [hunk 4]
```diff
@@ -51,13 +51,12 @@ class Engine < ::Rails::Engine
...
       end
 
-      menu :project_menu,
-           :bcf,
-           { controller: '/bcf/issues', action: :index },
-           caption: :'bcf.label_bcf',
-           param: :project_id,
-           icon: 'icon2 icon-backlogs',
-           badge: 'bcf.experimental_badge'
+      rename_menu_item :project_menu,
```
Comment (wielinde) [inline modules/bcf/lib/open_project/bcf/engine.rb:56]: nice! Maybe we could even extend it to take a different URL to call? It would be great if this menu item and the view "All open" opens the grid view.

File: lib/redmine/menu_manager/menu_item.rb [hunk 5]
```diff
@@ -88,6 +112,10 @@ def html_options(options = {})
   def add_condition(new_condition)
     raise ArgumentError, 'Condition needs to be callable' unless new_condition.respond_to?(:call)
     old_condition = @condition
-    @condition = -> (project) { old_condition.call(project) && new_condition.call(project) }
+    if old_condition.respond_to?(:call)
```
Comment (wielinde) [inline lib/redmine/menu_manager/menu_item.rb:127]: I am curious why this change became necessary here. How can it be that this worked before?

**Review by wielinde — CHANGES_REQUESTED** [review]
Hey @HDinger thank you for that great PR. I especially like how you decide on what is core and what is module. And sure, that separation is tricky in the frontend. Well done.

My comments are pretty small. So just have a quick look at them and then feel free to merge.

File: lib/redmine/menu_manager/menu_item.rb (hunk 5 above)
Comment (HDinger) [reply lib/redmine/menu_manager/menu_item.rb:127]: It did not work before, when there was initially no condition given for the element. Since the method is not used very often, and those elements who used it have a condition, we didn't notice it.
